bottlecap/src/logs/flusher.rs

```diff
@@ -55,73 +64,116 @@ impl Flusher {
             if let Err(e) = result {
-                debug!("Failed to send logs: {}", e);
+                debug!("Failed to join task: {}", e);
+                continue;
+            }
+
+            // At this point we know the task completed successfully,
+            // but the send operation itself may have failed
+            if let Err(e) = result {
+                if let Some(failed_req_err) = e.downcast_ref::<FailedRequestError>() {
+                    // Clone the request from our custom error
+                    failed_requests.push(
+                        failed_req_err
+                            .request
+                            .try_clone()
+                            .expect("should be able to clone request"),
+                    );
+                    debug!("Failed to send logs after retries, will retry later");
+                } else {
+                    debug!("Failed to send logs: {}", e);
+                }
             }
         }
+        failed_requests
     }
 
     fn create_request(&self, data: Vec<u8>) -> reqwest::RequestBuilder {
         let url = format!("{}/api/v2/logs", self.config.logs_config_logs_dd_url);
         let body = self.compress(data);
         self.client
             .post(&url)
+            .timeout(std::time::Duration::from_secs(self.config.flush_timeout))
             .headers(self.headers.clone())
             .body(body)
     }
 
     async fn send(req: reqwest::RequestBuilder) -> Result<(), Box<dyn Error + Send>> {
-        let time = Instant::now();
-        let resp = req.send().await;
-        let elapsed = time.elapsed();
-
-        match resp {
-            Ok(resp) => {
-                let status = resp.status();
-                _ = resp.text().await;
-                if status != 202 {
-                    debug!(
-                        "Failed to send logs to datadog after {}ms: {}",
-                        elapsed.as_millis(),
-                        status
-                    );
-                }
-            }
-            Err(e) => {
-                error!(
-                    "Failed to send logs to datadog after {}ms: {}",
-                    elapsed.as_millis(),
-                    e
-                );
+        let mut attempts = 0;
+
+        loop {
+            let time = Instant::now();
+            attempts += 1;
+            let Some(cloned_req) = req.try_clone() else {
+                return Err(Box::new(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "can't clone",
+                )));
+            };
+            let resp = cloned_req.send().await;
+            let elapsed = time.elapsed();
 
-                return Err(Box::new(e));
+            match resp {
+                Ok(resp) => {
+                    let status = resp.status();
+                    _ = resp.text().await;
+                    if status == 202 {
+                        return Ok(());
```

Comment:
Wondering if we had the `status != 202` as that because there would be no error but the request was not successful, would it make sense to debug here?

Not sure TBH

Reply:
The common case here would be someone temporarily disabling an API key and then re-enabling it (eg: brown out test), so we'd retry in a 403 event. Any _successful_ request (eg: any status code at all), will hit this block. So I'm not sure we want to debug every case.

An argument could be made for not retrying in the case of a 403, but in the aforementioned brown-out case it may be nice.